### ISSUE TITLE
Build Python polars wheels with PGO

### ISSUE THREAD

**Issue by messense** [pola-rs/polars#9702]
### Problem description

It seems that [PGO](https://doc.rust-lang.org/rustc/profile-guided-optimization.html) is a success for [pydantic-core](https://github.com/pydantic/pydantic-core/pull/678#issuecomment-1600427788), I wonder whether it will speedup py-polars.

**Comment by ritchie46** [comment]
Yes, I'd be very interested in that. I think including the db-benchmark and the tpch benchmark code in the guided data seem to me to be interesting candidates.

Tests would be easiest. Not any experience with setting it up though, so any help on this is appreciated.

**Comment by messense** [comment]
@ritchie46 Is there any easy way to compare two benchmark results? I'd like to confirm that it actually improves performance before putting too much effort.

**Comment by ritchie46** [comment]
The easiest start are the TPCH benchmarks: https://github.com/pola-rs/tpch/tree/main/polars_queries

The repo has a `Makefile` that creates the dataset. Running this provides a timing table and some plot utils. Now that I think of it, that's also good PGO input.

**Comment by jmakov** [comment]
> 

From some of the blog posts it looks like +15% can be done. That's the difference a CPU makes from one generation to another.

**Comment by FilipAndersson245** [comment]
Any update on this? PGO is a very interesting way to improve performance with the only cost is deployment and build complexity.
I can recommend looking at https://github.com/Kobzol/cargo-pgo for inspiration on how to run it.

**Comment by alonme** [comment]
I started taking a look at that today.

It seemed that using the flow documented in https://doc.rust-lang.org/rustc/profile-guided-optimization.html  - and when benchmarking the PGO'd version after that  - i didn't any change in performance.

I suspect that the issue is that only the `py-polars` crate was actually compiled to generate the profiling data and then optimized.

Currently trying to make it work on all the crates

**Comment by FilipAndersson245** [comment]
I have only used Kobzols Cargo-pgo tool so i have less experience with the raw pgo scripts. @zamazan4ik the go to guy with everything pgo.

**Comment by alonme** [comment]
After Compiling with global flags e.g. `export RUSTFLAGS="-Cprofile-generate=/tmp/pgo-data` and then `export RUSTFLAGS=-Cprofile-use=/tmp/pgo-data/merged.profdata` errors when compiling the second one with

```module flag identifiers must be unique (or of 'require' type)
!"CG Profile"
rustc-LLVM ERROR: Broken module found, compilation aborted!
error: could not compile `py-polars` (lib)
```

which is referenced in the following issues:
https://github.com/rust-lang/rust/issues/115344
https://github.com/rust-lang/rust/issues/117220


I will try to check if a different LTO setting will make it compile, and check the effect

**Comment by alonme** [comment]
`lto = "thin"` compiles successfully 👍

**Comment by alonme** [comment]
> After Compiling with global flags e.g. `export RUSTFLAGS="-Cprofile-generate=/tmp/pgo-data` and then `export RUSTFLAGS=-Cprofile-use=/tmp/pgo-data/merged.profdata` errors when compiling the second one with
> 
> ```
> !"CG Profile"
> rustc-LLVM ERROR: Broken module found, compilation aborted!
> error: could not compile `py-polars` (lib)
> ```
> 
> which is referenced in the following issues: [rust-lang/rust#115344](https://github.com/rust-lang/rust/issues/115344) [rust-lang/rust#117220](https://github.com/rust-lang/rust/issues/117220)
> 
> I will try to check if a different LTO setting will make it compile, and check the effect

these rust issues have been resolved - would probably be a good idea to try again
